### PR TITLE
Add chapter on NLP #137

### PR DESCRIPTION
**Pull request by j-lum**
There's quite a bit of content that I had to prune out because it's quite challenging to explain in simple terms. 

Let me know if there's anything that's still hard to follow

### CODE REVIEW

**Review by damithc — COMMENTED** [review]
Some quick comments.
* The depth and length seem good.
* Add link to the main list of chapters
* Cite source if using images from elsewhere

Go ahead and get some peer reviews.

File: contents/naturalLanguage/nl.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
+
+![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>
+From the parts-of-speech tags, a tree-structured representation of grammar can be produced.
+This tree is the result of a _constituency parse_, an task of recognizing a sentence and assigning a syntactic structure to it.
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: `an task` -> `a task`

File: contents/naturalLanguage/nl.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+### Syntax
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
+
+![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: Could the blue borders be removed, or surround the entire image?

File: contents/naturalLanguage/nl.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+In order to understand the meaning of the sentence, a few questions need to be answered:
+
+* Who is 'I'?
+* What is 'an article'? 
+* How are the two subjects related?
+
+An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.
+
+![Relation tuple](openIE.png "Constituency Parse") <br/>
+From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What the articles that I have written".
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: is the structure in `What the articles that I have written` intentional?

File: contents/naturalLanguage/nl.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+This technique, Open Information Extraction, is utilized by IBM's Watson to defeat human players in the American game show Jeopardy!<sup>[8](#footnote8)</sup>
+
+### Relations 
+Relations is crucial to the understanding of natural languages.
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: `are`

File: contents/naturalLanguage/nl.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.
+
+### Applications of NLP
+
+#### Input methods editors, e.g. Chinese, Japanese and Korean
+Modern input method editors(IMEs) do more than simply translating input to output in the respective language.
+By tackling the syntactic and relational elements of natural language, IMEs provide autocomplete functionality by suggesting words that are related or occur together frequently. 
+
+For example, typing in "*beijing*" in Moon IME, a Chinese IME, will bring up "*olympics*" as a possible suggestion<sup>[12](#footnote12)</sup>.
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: ~in~

File: contents/naturalLanguage/nl.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
+Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special neural network architectures to model sentences.  
+
+#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
+Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 
+For example, the two possible input from users regarding a replacement of digital banking token:
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: `of a`

File: contents/naturalLanguage/nl.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
+Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 
+For example, the two possible input from users regarding a replacement of digital banking token: 
+
+> * I want to replace my token.
+> * My old token is broken. 
+
+![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>
+
+can be mapped back to the same intent to which a predefined response can be given.
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md:152]: Could move this above the second image

File: contents/naturalLanguage/nl.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>
+
+can be mapped back to the same intent to which a predefined response can be given. 
+
+
+## What's next
+1. Get an intuition for the English language
+
+    The Language Instinct by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics.
```
Comment (ang-zeyu) [inline contents/naturalLanguage/nl.md]: Missing link here

**Review by ang-zeyu — COMMENTED** [review]
Really succinct and easy to follow 👍 
Just presentation / language nits:

File: contents/naturalLanguage/nl.md (hunk 3 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Added quotes around the 'I' similar to the previous examples for emphasis.

**Comment by j-lum** [comment]
@ang-zeyu Thanks for the through review!

File: contents/naturalLanguage/nl.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+* [Core problems in NLP](#core-problems-in-nlp)
+   * [Syntax](#syntax)
+   * [Semantics](#semantics)
+   * [Relations](#relations)
+* [Applications of NLP](#applications-of-nlp)
+* [What's next](#whats-next)
+* [References](#references)
+
+
+## What is NLP
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md]: I think we're standardising questions to have question marks so it should be `What is NLP?`

File: contents/naturalLanguage/nl.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+## What's next
+1. Get an intuition for the English language
+
+    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
+    This is invaluable especially for those who grew up with English as a first language.
+
+1. Visually explore the themes present in NLP
+    
+    Standford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md]: Stanford's

File: contents/naturalLanguage/nl.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
+Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
+spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
+automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;
+machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 
+ 
+NLP draws from many other fields of science, from formal linguistics to statistics. 
+The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.
+
+Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes.
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md]: Might be good to state what the three themes are, i.e. "most of them can be generalized to three themes: Syntax, Semantics and Relations, which are discussed in the next section"

File: contents/naturalLanguage/nl.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+## Themes in NLP
+
+### Syntax
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md]: Might be good to increase the image size and add a subheader for an image using the 'pic' syntax below.

```xml
<pic src="partsOfSpeech.png" alt="partsOfSpeech" width="70%">

  <sub> _Figure 1. Parts of Speech_ </sub>

</pic>
```

It would also be more aesthetic to centre the image.

File: contents/naturalLanguage/nl.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
+
+![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>
+From the parts-of-speech tags, a tree-structured representation of grammar can be produced.
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md:61]: Likewise, can add a subheader here

File: contents/naturalLanguage/nl.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+In order to understand the meaning of the sentence, a few questions need to be answered:
+
+* Who is 'I'?
+* What is 'an article'? 
+* How are the two subjects related?
+
+An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.
+
+![Relation tuple](openIE.png "Constituency Parse") <br/>
+From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What the articles that 'I' have written".
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md]: "What are the articles that 'I' have written?"

File: contents/naturalLanguage/nl.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+
+A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
+This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.
+
+Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.
+
+### Applications of NLP
+
+#### Input methods editors, e.g. Chinese, Japanese and Korean
+Modern input method editors(IMEs) do more than simply translating input to output in the respective language.
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md:128]: Might be good to elaborate on what Method Input Editors are, since it's introduced here for the first time

File: contents/naturalLanguage/nl.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+> * I want to replace my token.
+> * My old token is broken. 
+
+can be mapped back to the same intent to which a predefined response can be given. 
+
+![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>
+
+## What's next
+1. Get an intuition for the English language
```
Comment (tiuweehan) [inline contents/naturalLanguage/nl.md]: Might be good to bold each of the points to make it stand out more

**Review by tiuweehan — APPROVED** [review]
Interesting read!

**Comment by j-lum** [comment]
Something interesting is that the <markdown> tag ignores all the formatting that it is nested in.
e.g 
`<sub><markdown>text</markdown></sub>` renders without inheriting any styles from `<sub>`. Not sure if this is intended behavior.

File: contents/naturalLanguage/nl.md [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+</frontmatter>
+
+<div class="website-content">
+
+{{ booktitle | safe }}
+
+# Introduction to Natural Language Processing (NLP)
+Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)
+
+* [What is NLP?](#what-is-nlp)
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md:18]: Can wrap this TOC in a 
```
<box id="article-toc">`
... 
</box>
```
, this will ensure that the TOC is responsive with the screen re-sizing

File: contents/naturalLanguage/nl.md [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
+Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
+spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
+automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;
+machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 
+ 
+NLP draws from many other fields of science, from formal linguistics to statistics. 
+The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.
+
+Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes: syntax, semantics and relations.
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md]: perhaps can italicise the syntax, semantics, relations and allow them to be clickable that navigates to their corresponding sections?

File: contents/naturalLanguage/nl.md [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+
+<div class="website-content">
+
+{{ booktitle | safe }}
+
+# Introduction to Natural Language Processing (NLP)
+Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)
+
+* [What is NLP?](#what-is-nlp)
+* [Core problems in NLP](#core-problems-in-nlp)
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md]: The section Core problems in NLP does not match with the Themes in NLP header

File: contents/naturalLanguage/nl.md [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+<p align="center">
+    <pic src="partsOfSpeech.png" alt="partsOfSpeech" width="90%">
+      <sub><i>Figure 1. Parts of Speech</i></sub>
+    </pic>
+</p>
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md]: in the preview, the footnote 5 seems to not work
Screen Shot 2020-03-27 at 12 44 54 PM

File: contents/naturalLanguage/nl.md [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
+Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
+Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special neural network architectures to model sentences.  
+
+#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
+Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>.
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md]: mistake of putting 14 instead of 15

**Review by jamessspanggg — APPROVED** [review]
Good read, planning to take NLP next semester so this was a good brief introduction. Some comments:

File: contents/naturalLanguage/nl.md [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+   * [Syntax](#syntax)
+   * [Semantics](#semantics)
+   * [Relations](#relations)
+* [Applications of NLP](#applications-of-nlp)
+* [What's next](#whats-next)
+* [References](#references)
+
+
+## What is NLP?
+Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
```
Comment (Tejas2805) [inline contents/naturalLanguage/nl.md:29]:
```suggestion
Natural Language Processing (NLP) is the set of methods that makes human language accessible to computers.
```

File: contents/naturalLanguage/nl.md [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+
+can be mapped back to the same intent to which a predefined response can be given. 
+
+<p align="center">
+    <pic src="dbsChatbot.png" alt="DBS chatbot responding to two different messages with a similar intent" width="90%">
+      <sub><i>Figure 6. DBS chatbot responding to two different messages with a similar intent</i></sub>
+    </pic>
+</p>
+
+## What's next
```
Comment (Tejas2805) [inline contents/naturalLanguage/nl.md]:
```suggestion
## What's Next?
```

File: contents/naturalLanguage/nl.md [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+      <sub><i>Figure 5. word2vec visualization for antonyms</i></sub>
+    </pic>
+</p>
+
+A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
+This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.
+
+Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.
+
+### Applications of NLP
```
Comment (Tejas2805) [inline contents/naturalLanguage/nl.md:125]: All headings in this section are quite lengthy. Can you prune them?

**Review by Tejas2805 — COMMENTED** [review]
We are aiming for consistency across the book by using `titleCase` for the headings. Have left a few comments and suggestions accordingly

File: contents/naturalLanguage/nl.md (hunk 22 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md:29]: "for making" has the implication that the methods were tailored for the said purpose.
"that makes" loses a bit of that meaning IMO.

File: contents/naturalLanguage/nl.md [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes: [_syntax_](#syntax), [_semantics_](#semantics) and [_relations_](#relations).
+
+## Themes in NLP
+
+### Syntax
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+<p align="center">
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md:47]: can use the md syntax `<center></center>` to center the images

File: contents/naturalLanguage/nl.md [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+    <pic src="partsOfSpeech.png" alt="partsOfSpeech" width="90%">
+      <sub><i>Figure 1. Parts of Speech</i></sub>
+    </pic>
+</p>
+
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[6](#footnote5)</sup>.
```
Comment (jamessspanggg) [inline contents/naturalLanguage/nl.md]: should this be `[5]` instead?

**Review by jamessspanggg — COMMENTED** [review]
Just some small nits

File: contents/naturalLanguage/nl.md [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+* [What's Next](#whats-next)
+* [References](#references)
+</box>
+
+## What is NLP?
+Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
+Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
+spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
+automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;
+machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md:33]: These references feel a bit random. Usually, the reference should point to the source of the fact given in the sentence. If they are simply examples, best to give them specifically as examples.

File: contents/naturalLanguage/nl.md [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+ 
+NLP draws from many other fields of science, from formal linguistics to statistics. 
+The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.
+
+Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes: [_syntax_](#syntax), [_semantics_](#semantics) and [_relations_](#relations).
+
+## Themes in NLP
+
+### Syntax
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: Abrupt usage of ML here. The relation is not explained above.

File: contents/naturalLanguage/nl.md [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+    <pic src="constituencyParse.png" alt="Constituency Parse" width="90%">
+      <sub><i>Figure 2. Constituency Parse</i></sub>
+    </pic>
+</p>
+
+From the parts-of-speech tags, a tree-structured representation of grammar can be produced.
+This tree is the result of a _constituency parse_, a task of recognizing a sentence and assigning a syntactic structure to it.
+From the tree we can see that the complex phrase 'word-level parts-of-speech' is simply a _noun phrase_.
+Incidentally, this is how Microsoft Word checks for grammatical errors<sup>[6](#footnote6)</sup>.
+Any tree that cannot be parsed may be grammatically incorrect or difficult to understand.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md:65]: Should have used an unrelated sentence in the example so as not to add unnecessary complications (never go meta with examples as it always make the reader's job harder). Was this example taken from elsewhere? if so, cite the source. If you designed the example, consider using a different sentence.

File: contents/naturalLanguage/nl.md [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.
+
+<p align="center">
+    <pic src="openIE.png" alt="Relation tuple" width="90%">
+      <sub><i>Figure 3. Relation tuple</i></sub>
+    </pic>
+</p>
+
+From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What are the articles that 'I' have written".
+This technique, Open Information Extraction, is utilized by IBM's Watson to defeat human players in the American game show Jeopardy!<sup>[8](#footnote8)</sup>
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: _Watson uses this technique to defeat human player_ seems to convey a misleading message. Isn't it _Watson, who is able to defeat human players, uses this technique_? Even then, I'm not sure there is any direct link between the technique and defeating the human player.

File: contents/naturalLanguage/nl.md [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+Relations are crucial to the understanding of natural languages.
+Consider the example:
+
+   > Geoffrey bought Mary a ring. They have been dating for months.
+
+What is the event that prompted Geoffrey to buy a ring? 
+The word _proposal_, while not stated in the example, implicitly links the two events together.
+While humans are able to draw the connections instinctually, it is not easy to formalize this process computationally.
+
+There are attempts to create semantic _ontologies_ for natural language, groupings of words by their meanings.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: Unexplained term: ontology

Comment (damithc) [reply contents/naturalLanguage/nl.md]: Explanation of new terms can be given as tooltips, BTW.

File: contents/naturalLanguage/nl.md [hunk 32]
```diff
@@ -0,0 +1,197 @@
...
+      <sub><i>Figure 5. word2vec visualization for antonyms</i></sub>
+    </pic>
+</p>
+
+A simplified visualization of words in vector space shows that not only are words related in meaning close together, their antonyms are clustered together as well<sup>[11](#footnote11)</sup>. 
+This implies that word vectors can be somewhat meaningfully combined by just using simple vector addition.
+
+Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.
+
+### Applications of NLP
```
Comment (damithc) [inline contents/naturalLanguage/nl.md:125]: This heading should be level 2?

File: contents/naturalLanguage/nl.md [hunk 33]
```diff
@@ -0,0 +1,197 @@
...
+Search terms on Google are grouped together by semantic similarity in real time at volume to identify trends worldwide. 
+This is achieved with techniques similar to **word2vec**. 
+
+#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
+Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
+Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special neural network architectures to model sentences.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: _an_ algorithm?

Comment (damithc) [inline contents/naturalLanguage/nl.md]: Unexplained term: neural network

File: contents/naturalLanguage/nl.md [hunk 34]
```diff
@@ -0,0 +1,197 @@
...
+1. **Read research papers to gain a deeper understanding of the techniques employed**
+
+    Stay up to date with research through portals like [arxiv sanity preserver](https://www.arxiv-sanity.com/search?q=nlp).
+    
+    
+## References
+
+<a name="footnote1">[1]</a>: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6562150/ <br/>
+<a name="footnote2">[2]</a>: https://azure.microsoft.com/en-us/services/bot-service/ <br/>
+<a name="footnote3">[3]</a>: https://github.com/google/seq2seq <br/>
```
Comment (damithc) [inline contents/naturalLanguage/nl.md:187]: Same as 14?

File: contents/naturalLanguage/nl.md [hunk 35]
```diff
@@ -0,0 +1,197 @@
...
+    <pic src="dbsChatbot.png" alt="DBS chatbot responding to two different messages with a similar intent" width="90%">
+      <sub><i>Figure 6. DBS chatbot responding to two different messages with a similar intent</i></sub>
+    </pic>
+</p>
+
+## What's Next
+1. **Get an intuition for the English language**
+
+    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
+    This is invaluable especially for those who grew up with English as a first language.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: The reason for this is not obvious.

File: contents/naturalLanguage/nl.md [hunk 36]
```diff
@@ -0,0 +1,197 @@
...
+can be mapped back to the same intent to which a predefined response can be given. 
+
+<p align="center">
+    <pic src="dbsChatbot.png" alt="DBS chatbot responding to two different messages with a similar intent" width="90%">
+      <sub><i>Figure 6. DBS chatbot responding to two different messages with a similar intent</i></sub>
+    </pic>
+</p>
+
+## What's Next
+1. **Get an intuition for the English language**
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: So, NLP is tied to English?

File: contents/naturalLanguage/nl.md [hunk 37]
```diff
@@ -0,0 +1,197 @@
...
+    Explore how sentences are annotated with *parts-of-speech* tags and see the output of a constituency parse.
+    
+1. **Experiment with NLP libraries**
+
+    The aforementioned [CoreNLP](https://stanfordnlp.github.io/CoreNLP/) library is a good starting point for developers comfortable in Java.
+    For a simple and productive experience, [spaCy](https://spacy.io/) is a Python library suitable for experimentation and rapid prototyping.
+    
+1. **Read research papers to gain a deeper understanding of the techniques employed**
+
+    Stay up to date with research through portals like [arxiv sanity preserver](https://www.arxiv-sanity.com/search?q=nlp).
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: I'm not sure we should pretend to guide the reader to _self-learn_  NLP. Is that even practical? Perhaps just point the reader to a good online NLP course?

**Review by damithc — CHANGES_REQUESTED** [review]
Good start. Some comments added.

File: contents/naturalLanguage/nl.md (hunk 29 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md:65]: Changed it to a simpler sentence.

File: contents/naturalLanguage/nl.md (hunk 28 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Rephrased to match the definition given in the intro.

File: contents/naturalLanguage/nl.md (hunk 33 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Added a tooltip.

File: contents/naturalLanguage/nl.md (hunk 31 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Added a tooltip.

File: contents/naturalLanguage/nl.md (hunk 30 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Changed the example to some other downstream application.

File: contents/naturalLanguage/nl.md (hunk 35 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Rephrased the point to be multilingual.

File: contents/naturalLanguage/nl.md (hunk 36 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Rephrased the point to include linguistics in general.

File: contents/naturalLanguage/nl.md (hunk 37 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md]: Included a reference to CS224n by Stanford.

File: contents/naturalLanguage/nl.md (hunk 34 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md:187]: Removed.

File: contents/naturalLanguage/nl.md (hunk 22 above)
Comment (Tejas2805) [reply contents/naturalLanguage/nl.md:29]: Ah I got your point. I just feel the sentence is a bit weirdly framed.

File: contents/naturalLanguage/nl.md [hunk 38]
```diff
@@ -0,0 +1,198 @@
...
+Search terms on Google are grouped together by semantic similarity in real time at volume to identify trends worldwide. 
+This is achieved with techniques similar to **word2vec**. 
+
+#### Aggregation and clustering of documents, e.g. Cambridge Analytica, Google News
+Techniques like **doc2vec**<sup>[13](#footnote13)</sup> build upon **word2vec** and provide a representation of paragraphs in vector space.
+Similar to **word2vec**, this enables classification of documents and paves the road to powerful information retrieval techniques. 
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[3](#footnote3)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special <tooltip content="artifical neural networks a computer system made up of a number of simple, highly connected processing elements which translate information to some form of desired output"><i>artificial neural network</i></tooltip>   architectures to model sentences.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md]: >artifical neural networks a computer system made up ...

Something missing in that phrase.

File: contents/naturalLanguage/nl.md [hunk 39]
```diff
@@ -0,0 +1,198 @@
...
+
+## What's Next
+1. **Get an intuition for linguistics**
+
+    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
+    Learn about how to deconstruct language into formal structures for critical analysis through engaging examples.
+    
+1. **Visually explore the themes present in NLP**
+    
+    Stanford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.
```
Comment (damithc) [inline contents/naturalLanguage/nl.md:168]: I couldn't get this to run though

**Review by damithc — CHANGES_REQUESTED** [review]
Just a cosmetic nit

File: contents/naturalLanguage/nl.md (hunk 39 above)
Comment (j-lum) [reply contents/naturalLanguage/nl.md:168]: It can be a bit sluggish to load at times depending on the load of the server.
